### PR TITLE
feat(iroh-net): persist known peer info

### PR DESCRIPTION
**Pull request by divagant-martian**
## Description

Rough list of changes
- add an option to the `MagicEndpointBuilder` to set the path in which peers are saved. This is used to read peers on start up and store them on an interval and on shutdown.
- add `KnownPeers` to ease serializing and deserializing peer info
- move the logic of `add_known_addr` to the `PeerMap` to facilitate instantiating from a known peer list
- Add an `IrohPath` for the peers.
- Add an option to the `Node`s builder to set the path, passed down to the endpoint.
- For serialization it makes sense to have an object modelling the peer addressing info. This is the `AddrInfo`, which now replaces the custom `IrohInfo`

## Change checklist

- [x] Self-review.
- [x] Documentation updates if relevant.
- [x] Tests if relevant.

### CODE REVIEW

File: iroh-gossip/src/net.rs [hunk 1]
```diff
@@ -1,18 +1,14 @@
...
-use std::{
-    collections::HashMap, fmt, future::Future, net::SocketAddr, sync::Arc, task::Poll,
-    time::Instant,
-};
-
 use anyhow::{anyhow, Context};
 use bytes::{Bytes, BytesMut};
 use futures::{stream::Stream, FutureExt};
 use genawaiter::sync::{Co, Gen};
+use iroh_net::magic_endpoint::AddrInfo as IrohInfo;
```
Comment (Frando) [inline iroh-gossip/src/net.rs:7]: I'd say remove the `as IrohInfo` and just use `AddrInfo`

File: iroh-net/src/magic_endpoint.rs [hunk 2]
```diff
@@ -422,6 +438,21 @@ impl MagicEndpoint {
         Ok(())
     }
 
+    /// Get addressing information of this [`MagicEndpoint`].
+    pub async fn my_addr_info(&self) -> Result<AddrInfo> {
```
Comment (Frando) [inline iroh-net/src/magic_endpoint.rs:442]: I see that `magicsock/endpoint.rs` now has `addr_info` which returns exactly what we return here. Maybe route to that to have 1, not 2 async calls to the actor here. But can also be a followup.

**Review by Frando — COMMENTED** [review]
Looks good!

File: iroh-net/src/magic_endpoint.rs [hunk 3]
```diff
@@ -391,16 +408,15 @@ impl MagicEndpoint {
     /// If no UDP addresses are added, and the given `derp_region` cannot be dialed, it will error.
     pub async fn add_known_addrs(
         &self,
-        node_id: PublicKey,
+        peer: PublicKey,
```
Comment (dignifiedquire) [inline iroh-net/src/magic_endpoint.rs:411]: weren't we calling these `node_id`s? why the change back?

File: iroh-net/src/magicsock.rs [hunk 4]
```diff
@@ -400,6 +409,24 @@ impl MagicSock {
             .instrument(info_span!("derp.actor")),
         );
 
+        // load the peer data
+        let peer_map = match peers_path.as_ref() {
+            Some(path) if path.exists() => {
+                match PeerMap::load_from_file(path, actor_sender.clone()) {
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply iroh-net/src/magicsock.rs:415]: maybe just add a bulk add addr method?

File: iroh-net/src/magic_endpoint.rs [hunk 5]
```diff
@@ -17,18 +18,32 @@ use crate::{
...
-/// Builder for [MagicEndpoint]
+impl PartialEq for AddrInfo {
+    fn eq(&self, other: &Self) -> bool {
+        if self.derp_region != other.derp_region {
+            return false;
+        }
+        if self.endpoints.len() != other.endpoints.len() {
+            return false;
+        }
+        let mut self_eps = self.endpoints.clone();
```
[the remark this answers is not in the document]
Comment (divagant-martian) [reply iroh-net/src/magic_endpoint.rs:38]: I did that but changes felt intrusive to everything that calls connect. But agree it's a better option in general

File: iroh-net/src/magicsock/endpoint.rs [hunk 6]
```diff
@@ -1112,6 +1193,15 @@ impl PeerMap {
         self.by_ip_port.insert(*ipp, id);
     }
 
+    /// Saves the known peer info to the given path, returning the number of peers persisted.
+    pub(super) fn save_to_file(&self, path: &Path) -> anyhow::Result<usize> {
+        let known_peers = self.known_peers();
+        let count = known_peers.len();
+        let serialized = postcard::to_stdvec(&known_peers)?;
+        std::fs::write(path, serialized)?;
```
Comment (dignifiedquire) [inline iroh-net/src/magicsock/endpoint.rs:1201]: should this be async?

File: iroh-net/src/magic_endpoint.rs (hunk 3 above)
Comment (divagant-martian) [reply iroh-net/src/magic_endpoint.rs:411]: I asked about this in #1487 
We are using peer_id through the gossip and sync code so at this point it feels more consistent than node to me. As long as we pick one I'm happy to oblige

File: iroh-net/src/magicsock/endpoint.rs [hunk 7]
```diff
@@ -1112,6 +1193,15 @@ impl PeerMap {
         self.by_ip_port.insert(*ipp, id);
     }
 
+    /// Saves the known peer info to the given path, returning the number of peers persisted.
+    pub(super) fn save_to_file(&self, path: &Path) -> anyhow::Result<usize> {
+        let known_peers = self.known_peers();
```
Comment (dignifiedquire) [inline iroh-net/src/magicsock/endpoint.rs:1198]: This allocates the whole thing in memory, effectively duplicating memory usage for this moment. I wonder if we could serialize it in an iterator/streaming fashion. eg sth like this

```rust
let mut file = std::fs::File::open(path)?;
file.write_all(HEADER)?;
let mut buffer = Vec::new();
for peer in &self.known_peers() {
  let ser = postcard::to_slice(peer, &mut buffer)?;
  file.write_all(ser)?;
}
file.flush()?;
```

Comment (dignifiedquire) [reply iroh-net/src/magicsock/endpoint.rs:1198]: (and make `self.known_peers()` return an iterator instead of a vec)

File: iroh/src/config.rs [hunk 8]
```diff
@@ -58,6 +60,7 @@ impl From<&IrohPaths> for &'static str {
             IrohPaths::BaoFlatStorePartial => "blobs-partial.v0",
             IrohPaths::DocsDatabase => "docs.redb",
             IrohPaths::Console => "console",
+            IrohPaths::PeerData => "peers.postcard",
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply iroh/src/config.rs:63]: I'd rather do sth like `peers.bin`, but more a personal preference. Maybe we should include a version number?

File: iroh-net/src/magic_endpoint.rs (hunk 3 above)
Comment (dignifiedquire) [reply iroh-net/src/magic_endpoint.rs:411]: I see, I don't mind either way, as long as we are somewhat consistent about it 😅

File: iroh-net/src/magicsock/endpoint.rs (hunk 6 above)
Comment (dignifiedquire) [reply iroh-net/src/magicsock/endpoint.rs:1201]: Also, we shouldn't directly write to the path, but rather to a `tmp` file (fine in the same location) and rename when we have successfully written, to avoid having errors during writing leaving corrupted files behind

File: iroh-net/src/magicsock.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (divagant-martian) [reply iroh-net/src/magicsock.rs:415]: the endpoint is passed because the value is stored for each endpoint, but it's not used and peers are not sent via this (or any channel) so I don't think there is possibility of deadlock. 

This seems dependent on trusting that the inner code does not use the channel, but I think the sync/non async coloring saves us here:  `load_from_file` is non-async, so any async call would need to be spawned outside this(effectively doing what you suggest). Any other non-async call of the `try_send` fashion would fail in that scenario. So I think this is safe

File: iroh/src/config.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (divagant-martian) [reply iroh/src/config.rs:63]: since we have already introduced strum, I'll move that test to iterate the values so that future paths are automatically added.

About the version number, what would be the purpose of this? is it for us to understand what info is in there? otherwise we don't store or read a peer data version that would help in an upgrade for example. Maybe we can tackle upgrading a peer data format version once it changes

the `.postcard` was more of a hunch I guess, I'm fine with any

File: iroh-gossip/src/net.rs (hunk 1 above)
Comment (divagant-martian) [reply iroh-gossip/src/net.rs]: done in [afa69d0](https://github.com/n0-computer/iroh/pull/1488/commits/afa69d0c4e051eac5a6f9ff20a8fc8df8ce7f4d7)

File: iroh-net/src/magic_endpoint.rs [hunk 9]
```diff
@@ -154,6 +172,14 @@ impl MagicEndpointBuilder {
         self
     }
 
+    /// Optionally set the path where peer info should be stored.
```
[the remark this answers is not in the document]
Comment (divagant-martian) [reply iroh-net/src/magic_endpoint.rs:175]: done in d036c3c

File: iroh-net/src/magicsock/endpoint.rs (hunk 6 above)
Comment (divagant-martian) [reply iroh-net/src/magicsock/endpoint.rs:1201]: didn't get to do this but will look into it

File: iroh-net/src/magicsock/endpoint.rs (hunk 7 above)
Comment (divagant-martian) [reply iroh-net/src/magicsock/endpoint.rs:1198]: didn't get to do this but will look into it

File: iroh/src/config.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (divagant-martian) [reply iroh/src/config.rs:63]: this is done in https://github.com/n0-computer/iroh/pull/1507

File: iroh-net/src/magicsock/endpoint.rs [hunk 10]
```diff
@@ -1018,6 +1043,53 @@ pub(super) struct PeerMap {
...
+        })
+    }
+
+    /// Create a new [`PeerMap`] from data stored in `path`.
+    pub fn load_from_file(
+        path: impl AsRef<Path>,
+        msock_sender: mpsc::Sender<ActorMessage>,
+    ) -> anyhow::Result<Self> {
+        let mut me = PeerMap::default();
+        let contents = std::fs::read(path)?;
```
Comment (Frando) [inline iroh-net/src/magicsock/endpoint.rs:1061]: We could read in chunks here. However the file would need to get *really* big for this to be an issue. A `PeerAddr` will take less than 100 bytes even with a few socket addresses. So if you store 10.000 peers that'd do a ~1MB allocation here.
I think it's fine to keep as-is now, and we could move to chunked reading easily in a follow-up.

**Review by Frando — APPROVED** [review]
LGTM! Let's get this in :+1: 

We might want to add a full test with a "recovering" endpoint (shutdown, restart, reconnect) to test the save-on-shutdown and read-on-bin, but that can also be a followup I'd say.
